src/octree/generation.rs

```diff
@@ -66,35 +67,48 @@ fn split<P>(
     stream: P,
 ) -> (Vec<octree::NodeId>, Vec<octree::NodeId>)
 where
-    P: Iterator<Item = Point>,
+    P: Iterator<Item = PointsBatch> + NumberOfPoints,
 {
     let mut children: Vec<Option<RawNodeWriter>> =
         vec![None, None, None, None, None, None, None, None];
-    match stream.size_hint() {
-        (_, Some(size)) => println!(
+    match stream.num_points() {
+        Some(size) => println!(
             "Splitting {} which has {} points ({:.2}x MAX_POINTS_PER_NODE).",
             node_id,
             size,
             size as f64 / MAX_POINTS_PER_NODE as f64
         ),
-        (_, None) => println!(
+        None => println!(
             "Splitting {} which has an unknown number of points.",
             node_id
         ),
     };
 
     let bounding_cube = node_id.find_bounding_cube(&Cube::bounding(&octree_meta.bounding_box));
-    stream.for_each(|p| {
-        let child_index = octree::ChildIndex::from_bounding_cube(&bounding_cube, &p.position);
-        let array_index = child_index.as_u8() as usize;
-        if children[array_index].is_none() {
-            children[array_index] = Some(RawNodeWriter::from_data_provider(
-                octree_data_provider,
-                octree_meta,
-                &node_id.get_child_id(child_index),
-            ));
+    stream.for_each(|batch| {
+        let child_indices: Vec<_> = batch
+            .position
+            .iter()
+            .map(|p| octree::ChildIndex::from_bounding_cube(&bounding_cube, &p))
+            .collect();
+        for (array_index, child_writer) in children.iter_mut().enumerate() {
+            let mut child_batch = batch.clone();
+            let keep: Vec<_> = child_indices
+                .iter()
+                .map(|i| i.as_u8() == array_index as u8)
+                .collect();
+            child_batch.retain(&keep);
```

Comment:
Could be made a little more efficient by making retain take an iterator, right?

Reply:
But not needed unless you think this function is particularly hot

Reply:
Or even better, do not clone and remove, but start from an empty point_batch and add points

Reply:
You are right, an iterator would be a little more efficient. However, I wanted to `assert` equal lengths of the keep vector and the internal vector, so for this we would need an `ExactSizeIterator`, which I didn't like as a prerequisite...
I actually also tested an implementation without `retain` just consisting of using `split_off` and `append`, which was much slower than cloning and removing. If you have a better idea, I'll be happy to test it, otherwise, I would like to keep it the way it is.